README.md
# whatsapp-converter

## Introduction
Use the whatsapp-converter to convert your exported WhatsApp chat to a CSV, XLS, XLSX or an ODS file.
1. The conversion is done locally. No data is shared with the internet!
2. The resultset is a CSV file which you can import into your favorite calculation application such as Excel or LibreOffice. You can also directly export to a spreadsheet file and do create a pivot table.

## Dependencies
If you want to install the dependencies manually, use pip. On some computers you might need to replace pip by pip3.
```shell
pip install tqdm pyexcel pyexcel-xlsxw pyexcel-ods3 xlwt
```

## Usage
Use the following command on the command line to get an overview of the available arguments.
```shell
python whatsapp-converter.py -h
```

### Convert WhatsApp TXT to CSV (Default Use Case)
The easy start is just to provide the filename of the exported WhatsApp chat. The application will process the chat and create a CSV file named *resultset.csv*.
```shell
python whatsapp-converter.py whatsapp-export.txt
```

### Convert WhatsApp to Excel XLS or XLSX
If you want to export to an Excel spreadsheet file instead, make sure to add the filename at the end, e.g. *resultset.xls*.
```shell
python whatsapp-converter.py whatsapp-export.txt resultset.xls
```

or

```shell
python whatsapp-converter.py whatsapp-export.txt resultset.xlsx
```

### Convert WhatsApp to LibreOffice ODS
If you want to export to a LibreOffice spreadsheet file, make sure to add the filename at the end, e.g. *resultset.ods*.
Please note that the export takes some time. The script did not crash.
```shell
python whatsapp-converter.py whatsapp-export.txt resultset.ods
```

Here is the list of all available options:
```shell
whatsapp-converter.py [-h] [-v] [-d] filename [resultset.csv|resultset.xls]
```

## Commandline Arguments

#### Mandatory Arguments
```shell
filename        The WhatsApp file containing the exported chat
resultset       Filename of the resultset, default resultset.csv. Use .csv to write a comma separated file. Use .xls to write to an Excel spreadsheet file
```

#### Optional Arguments
```shell
  -h, --help      show this help message and exit
  -n  --newline   create a new line (same date and time) in the resultset for every multiline chat message
  -v, --verbose   increase output verbosity
  -d, --debug     increase output verbosity to debug
```

##### Example Newline Argument

Consider this chat exported chat message
```shell
18.05.19, 11:50:00: Alan Smith: It starts here
This is a new line
```

By default the resultset will merge the two lines.
```shell
2019-05-18 11:50|2019-05-18|11:50|Alan Smith|It starts here This is a new line
```

If you start the application with the -n argument, a new line will be added and the date and time of the multiline message taken.
```shell
2019-05-18 11:50|2019-05-18|11:50|Alan Smith|It starts here
2019-05-18 11:50|2019-05-18|11:50|Alan Smith|This is a new line
```

## Conversion from to

Each line of the dataset will be converted to structured data which can be imported into your Excel or LibreOffice sheet.
```shell
21/8/19, 11:28:51 PM: Jon Doe: Waste out of wealth 😂
21.12.16, 23:29 - Alan Smith: Like a Rolex
30.11.18, 10:57 - Nachrichten in diesem Chat sowie Anrufe sind jetzt mit Ende-zu-Ende-Verschlüsselung geschützt. Tippe für mehr Infos.
30.11.18, 10:57 - Snoopy: Hallo :)
30.11.18, 10:58 - Snoopy: Jetzt bin ich gespannt auf deine Antwort ;)
1/26/19, 00:07 - Jon Doe: Sleep well
18.05.19, 11:50:00: Alan Smith: It starts here
This is a new line
```

The resultset file looks like this. In the default configuration new lines get the same date and time stamp and sender name as the previous line:
```shell
Date and Time|Date|Time|Name|Message
2019-8-21 23:28|2019-8-21|23:28|Jon Doe|Waste out of wealth 😂
2016-12-21 23:29|2016-12-21|23:29|Alan Smith|Like a Rolex
2018-11-30 10:57|2018-11-30|10:57|Snoopy|Hallo :)
2018-11-30 10:58|2018-11-30|10:58|Snoopy|Jetzt bin ich gespannt auf deine Antwort ;)
2019-01-26 00:07|2019-01-26|00:07|Jon Doe|Sleep well
2019-05-18 11:50|2019-05-18|11:50|Alan Smith|It starts here This is a new line
```


CHANGES.md
v0.6.4
Corrected small bug

v0.6.3
Improved the algorithmic efficiency

v0.6.2
Improved the algorithmic efficiency

v0.6.1
Newline command line argument fixed.

v0.6.0
Fixed a bug which did not allow to import and use the convert function in Python applications.

v0.5.7
Added 21/12/19 Date Format with square brackets and some date format error coverage

v0.5.3
Added 21/12/2019 Date Format with square brackets and am

v0.5.2
Corrected a small bug which prevented the colors.py to be loaded in main

v0.5.1
Implemented a switch to only show bash color codes for Linux.

v0.5.0
Added multiline support. Merge it or export it as a new line.

v0.4.5
Moved color definition into separate file

v0.4.4
Added support for dd/mm/yyyy, 23:59 - name: message format. Solved issue #5.

v0.4.3
Correct bug with square brackets and month/day/year. Solved parts of issue #4.

v0.4.2
Added support for another date and time format in square brackets. Solved parts of issue #4.

v0.4.1
Added support for date and time in square brackets. Solved parts of issue #4.

v0.4.0
Optimized code and added an Excel xlsx export and corrected the LibreOffce export

v0.3.8
Fixed an issue in the Excel export, introduced classes

v0.3.7
Fixed a bug in __main__.py which prevented the script from executing

v0.3.6
Refactored code after lint analysis

v0.3.5
Corrected small bug

v0.3.3
Removed unused variables and corrected export filename

v0.3.2
Corrected bug with UK date format

v0.3.1
Corrected missing dependency

v0.3.0
Added export to LibreOffice spreadsheet file

v0.2.3
Correct bug with UTF-8 import/export

v0.2.2
Added a progress bar

v0.2.1
Replaced old version of the README.md

v0.2.0
Working version to convert Whatsapp chat exports to csv files.


tests/test_parse.py
from unittest import TestCase
from whatsapp_converter import whatsapp_converter

class TestParse(TestCase):

    def test_is_string(self):
        # s=whatsapp_converter.parse("2/20/19, 10:14 PM - Messages to this group are now secured with end-to-end encryption. Tap for more info.", 1)
        s=whatsapp_converter.parse("23.12.18, 21:06 - Thomas: Auf alle Fälle. Lass uns schauen, dass wir uns nächstes Jahr mal treffen.", 1)
        self.assertEqual(s, 'FOO')

if __name__ == '__main__':
    unittest.main()


whatsapp_converter/tests/__main__.py
import os
import sys
import unittest

if __name__ == "__main__":
    # Get the current directory (where this file is located)
    current_dir = os.path.dirname(os.path.abspath(__file__))

    # Add the relative path to your project directory
    project_dir = os.path.join(current_dir, "..")  # Adjust ".." based on your project's directory structure
    sys.path.append(project_dir)

    # Import and run the tests
    from test_parse import TestParse

    test_suite = unittest.TestLoader().loadTestsFromTestCase(TestParse)
    unittest.TextTestRunner().run(test_suite)

whatsapp_converter/tests/test_parse.py
import unittest
import argparse  # Import the argparse module
from whatsapp_converter import parse


class TestParse(unittest.TestCase):
    def test_parse_with_valid_line(self):
        # Test parse function with a valid line
        line = "17/05/2021, 12:50 - Vijaya: Hey! This is Vijaya"
        local_args = argparse.Namespace(debug=False, verbose=False)
        result = parse(line, local_args)

        # Assert the output of the parse function
        self.assertEqual(result, ['new', '2021-05-17 12:50', '2021-05-17', '12:50', 'Vijaya', 'Hey! This is Vijaya'])

    def test_parse_with_valid_line(self):
        # Test parse function with a valid line
        line = "21/12/16, 10:50:12 PM: Neha Wipro: Awsum"
        local_args = argparse.Namespace(debug=False, verbose=False)
        result = parse(line, local_args)

        # Assert the output of the parse function
        self.assertEqual(result, ['new', '2016-12-21 22:50', '2016-12-21', '22:50', 'Neha Wipro', 'Awsum'])

    def test_parse_with_empty_line(self):
        # Test parse function with an empty line
        line = "\n"
        local_args = argparse.Namespace(debug=False, verbose=False)
        result = parse(line, local_args)

        # Assert the output of the parse function
        self.assertEqual(result, ['empty', '', '', '', '', ''])

if __name__ == '__main__':
    unittest.main()

whatsapp_converter/__main__.py
# __main__.py

# from importlib import resources  # Python 3.7+
# import sys

import os
import sys
import argparse
import platform
from typing import Dict

import whatsapp_converter
from whatsapp_converter import colors

# the path to the directory this file is in
_MODULE_PATH = os.path.dirname(__file__)

VERSION: Dict[str, str] = {}
with open(_MODULE_PATH + "/version.py", "r") as version_file:
    exec(version_file.read(), VERSION)

# ---------------------------------------------
# Supported filetypes for the resultset
SUPPORTED = (
    ".csv",
    ".xls",
    ".xlsx",
    ".ods"
)

def main():
    parser = argparse.ArgumentParser(prog='whatsapp-converter', description='Use whatsapp-converter to convert your exported WhatsApp chat to a CSV, ODS (LibreOffice) or XLS (Excel) file.', epilog='For reporting bugs or requesting features, please visit https://github.com/sandsturm/whatsapp-converter/ and create an issue')
    parser.add_argument('filename', metavar='filename', type=str, help='the WhatsApp file containing the exported chat')
    parser.add_argument('resultset', default='resultset.csv', nargs='?', help='filename of the resultset, default resultset.csv. Use .csv to write a comma separated file. Use .xls or .xlsx to write to an Excel spreadsheet file. Use .ods to write to a LibreOffice file.')
    parser.add_argument('-n', '--newline', help='create a new line (same date and time) in the resultset for every multline chat message', action='store_true')
    parser.add_argument('-v', '--verbose', help='increase output verbosity', action='store_true')
    parser.add_argument('-d', '--debug', help='increase output verbosity to debug', action='store_true')
    parser.add_argument('--version', action='version', version="v" + VERSION["VERSION"])

    # parser.add_argument("-nl", "--newline", help = "message across various lines is counted as a new message", action = "store_true")
    args = parser.parse_args()

    if not str( args.filename ):
        if( platform.system() == "Linux" ):
            print(f'{colors.BCOLORS["FAIL"]}ERROR: whatsapp-converter needs at least an import file.{colors.BCOLORS["ENDC"]}')
        else:
            print("ERROR: whatsapp-converter needs at least an import file.")
        sys.exit()

    if not str( args.resultset ).endswith( SUPPORTED ):
        if( platform.system() == "Linux" ):
            print(f'{colors.BCOLORS["FAIL"]}ERROR: whatsapp-converter only supports the following fileformats:{colors.BCOLORS["ENDC"]}')
        else:
            print("ERROR: whatsapp-converter only supports the following fileformats:")

        # TODO iterate over tuple
        print(SUPPORTED)
        sys.exit()

    whatsapp_converter.convert(filename=args.filename, resultset=args.resultset, newline=args.newline, verbose=args.verbose, debug=args.debug)


if __name__ == "__main__":
    main()


whatsapp_converter/whatsapp_converter.py
'''whatsapp-converter

Use this application to convert your exported WhatsApp chat to a CSV or Excel file.

1. The conversion is done locally. No data is shared with the internet!
2. The resultset is a CSV file which you can import into your favorite calculation application such as Excel or LibreOffice. You can also directly export to a XLS spreadsheet file.

'''

import io
import sys
import re
import datetime
from datetime import date
import pyexcel
from tqdm import tqdm


# ---------------------------------------------
# The main regex logic to identify the date and message
pattern_date = '^\[?((\d{1,2})([\/|\.])(\d{1,2})[\/|\.](\d{1,4}))\,?\ (\d{1,2}[:|.]\d{1,2})([:|.]\d{1,2})?\ ?(AM|PM|am|pm)?([\:\ |\ \-\ |\]\ ][^:])'
pattern = pattern_date + '(.+?)\:\ (.*)'

lastentry = {
    "lastdate": date.today(),
    "lasttime": date.today(),
    "lastname": ""
}

# ---------------------------------------------
# Define export file header
header = [
    "Date and Time",
    "Date",
    "Time",
    "Name",
    "Message"
]

date_formats = {
    "mdyS": False,
    "process_again": True,
    "data_format": "",
}


def check_monthfirst(dateformat):
    if dateformat[1:2] not in ('.', '/'):
        if int(dateformat[0:2]) > 12:
            return True
        else:
            return False


def parse(line, filename, resultset, newline, verbose, debug):
    '''Parse each line

    line = Single line to parse
    args = Arguments from the command line. In this case only verbose and debug switches.
    '''

    # ---------------------------------------------
    # The main regex logic to identify the date and message
    pattern_date = re.compile('^\[?((\d{1,2})([\/|\.])(\d{1,2})[\/|\.](\d{1,4}))\,?\ (\d{1,2}[:|.]\d{1,2})([:|.]\d{1,2})?\ ?(AM|PM|am|pm)?([\:\ |\ \-\ |\]\ ][^:])(.+?)\:\ (.*)')
    match = pattern_date.match(line)
    dataset = ['empty', '', '', '', '', '']

    # ---------------------------------------------
    # if verbose: print(prefix + line)
    # if verbose: print(BCOLORS.FAIL + prefix + line)
    # Identify the in the chat line

    if match:

        if match and match.group(10) !=  'M':

            date = datetime.datetime.strptime(match.group(1), date_formats["date_format"]).date()

            if match.group(8):
                time = datetime.datetime.strptime(match.group(6) + " " + match.group(8), "%I:%M %p").time()
            elif match.group(6).find(":") > 0:
                time = datetime.datetime.strptime(match.group(6), "%H:%M").time()
            elif match.group(6).find(".") > 0:
                time = datetime.datetime.strptime(match.group(6), "%H.%M").time()

            # ---------------------------------------------
            # Buffer date, time, name for next line messages
            lastentry["lastdate"] = date.strftime("%Y-%m-%d")
            lastentry["lasttime"] = time.strftime("%H:%M")
            lastentry["lastname"] = match.group(10)

            # ---------------------------------------------
            # Create the dataset for the new message
            dataset = ['new', date.strftime("%Y-%m-%d") + " " + time.strftime("%H:%M"), date.strftime("%Y-%m-%d"), time.strftime("%H:%M"), str(match.group(10)), match.group(11)]

            if verbose or debug:
                print(dataset)

    elif re.match(re.compile(r"^[\t ]*\n", re.VERBOSE), line):

        # ---------------------------------------------
        # Empty line
        if debug:
            print("Empty line removed")

    # ---------------------------------------------
    # Check if there is
    elif not re.match(re.compile(r"^\[?((\d{1,2})([\/|\.])(\d{1,2})[\/|\.](\d{1,4}))\,?\ (\d{1,2}[:|.]\d{1,2})(?:[:|.]\d{1,2})?\ ?(AM|PM|am|pm)?([\:\ |\ \-\ |\]\ ][^:])", re.VERBOSE), line):

        if debug:
            print("Appending line found")

        text = (re.match(re.compile(r"^(.*)", re.VERBOSE), line))

        # ---------------------------------------------
        # Create the dataset if commandline argument was to create a new line
        # TODO if (newline):
        # print(newline)
        if newline:
            dataset = ['new', str(lastentry["lastdate"]) + " " + str(lastentry["lasttime"]), str(lastentry["lastdate"]), str(lastentry["lasttime"]), lastentry["lastname"], text.group(0)]

        else:
            dataset = ['append', str(lastentry["lastdate"]) + " " + str(lastentry["lasttime"]), str(lastentry["lastdate"]), str(lastentry["lasttime"]), lastentry["lastname"], text.group(0)]

    return dataset


def convert(filename, resultset="resultset.csv", newline=False, verbose=False, debug=False):
    '''Convert the input file

    1. Open the file
    2. Loop to input line by line and send it to parser
    3. Close the file

    '''
    # ---------------------------------------------
    # Store the number of lines of the input file
    line_count = 0

    if verbose:
        print("Verbosity turned on")

    if debug:
        print("Debug turned on")

    try:
        with io.open(filename, "r", encoding="utf-8") as file:
            print("Reading import file")

            content = file.readlines()  # Read all lines at once
            line_count = len(content)  # Get the number of read lines
            confirmed_date = False  # Exit condition for For loop based on a date > 12

            for line in content:

                if confirmed_date == True:
                    break

                if not line.strip():  # Check if the line is empty or contains only whitespace
                    continue  # Skip the rest of the loop and move to the next iteration

                # ---------------------------------------------
                # Make the match, assign to the groups
                match = re.match(re.compile(pattern, re.VERBOSE), line)
                
                if match:

                    # ---------------------------------------------
                    # 21/12/19
                    if match.group(3) == '/' and (match.group(9) == ' -' or match.group(9) == '- ' or match.group(9) == ': '):
                        if debug:
                            print("Date format d/m/y")
                        if len(match.group(5)) == 4:
                            date_formats["date_format"] = "%d/%m/%Y"
                        else:
                            date_formats["date_format"] = "%d/%m/%y"

                    # ---------------------------------------------
                    # 12/21/19
                    elif match.group(3) == '/' and (match.group(9) == ' -' or match.group(9) == '- '):
                        if debug:
                            print("Date format m/d/y")
                        if len(match.group(5)) == 4 and match.group(7) is None:
                            date_formats["date_format"] = "%m/%d/%Y"
                        else:
                            date_formats["date_format"] = "%m/%d/%y"

                    # ---------------------------------------------
                    # 21/12/2019 with square brackets and am
                    elif match.group(3) == '/' and match.group(9) == '] ' and (match.group(8) == 'am' or match.group(8) == 'pm'):
                        if debug:
                            print("Date format d/m/Y with [square brackets] and am/pm")
                        date_formats["date_format"] = "%d/%m/%Y"

                    # ---------------------------------------------
                    # 12/21/2019 with square brackets
                    elif match.group(3) == '/' and match.group(9) == '] ':
                        if debug:
                            print("Date format m/d/Y with [square brackets]")
                        if check_monthfirst(match.group(1)) and not date_formats["mdyS"]:
                            date_formats["mdyS"] = True
                            date_formats["date_format"] = "%d/%m/%Y"
                        if date_formats["mdyS"]:
                            if len(match.group(5)) == 4:
                                date_formats["date_format"] = "%d/%m/%Y"
                            else:
                                date_formats["date_format"] = "%d/%m/%y"
                        else:
                            if len(match.group(5)) == 4:
                                date_formats["date_format"] = "%m/%d/%Y"
                            else:
                                date_formats["date_format"] = "%m/%d/%y"

                    # ---------------------------------------------
                    # 21.12.19
                    else:
                        if debug:
                            print("Date format d.m.y")
                        if len(match.group(5)) == 4:
                            date = datetime.datetime.strptime(match.group(1), "%d.%m.%Y").date()
                            date_formats["date_format"] = "%d.%m.%Y"
                        else:
                            date_formats["date_format"] = "%d.%m.%y"

    except IOError as e:
        print("File \"" + filename + "\" cannot be found.")
        sys.exit()

    print("Converting data now with data format " + date_formats["date_format"])

    # ---------------------------------------------
    # Count number of chatlines without  linesError: Excel 2003 only supports a maximum of 65535 lines.
    # Whatsapp-converter found more than 65535 lines for input which might lead to an error.\n
    counter = 0

    if debug:
        print('Open export file ' + resultset)

    # ---------------------------------------------
    # Select export formats
    if str(resultset).endswith('.xls'):
        if line_count > 65535:
            print(f'\n{BCOLORS["FAIL"]}Error: Excel 2003 only supports a maximum of 65535 lines. Whatsapp-converter found more than 65535 lines for input which might lead to an error.{BCOLORS["ENDC"]}\n')
            print("")
            sys.exit()#

    dataset = []
    dataset.append(header)

    # ---------------------------------------------
    # TODO Append line with buffer before writing
    # Show progress via tqdm
    for line in tqdm(content, total=line_count, ncols=120):
        if debug and line == '':
            print(line)

        buffer = parse(line, filename, resultset, newline, verbose, debug)

        if buffer[0] != 'empty':

            # ---------------------------------------------
            # Write to dataset
            if buffer[0] == 'new' or (newline and buffer[0] == 'append'):
                dataset.append(buffer[1:])
            # Default multiline appended to previous converted message
            else:
                dataset[-1][-1] = dataset[-1][-1] + " " + buffer[-1]

        # ---------------------------------------------
        # Print progress
        if line.strip():
            counter += 1

    print('Writing to ' + str(resultset))

    # ---------------------------------------------
    # Select export formats
    if str(resultset).endswith('.csv'):
        pyexcel.save_as(array=dataset, dest_file_name=str(resultset), dest_delimiter='|')

    elif str(resultset).endswith('.xls'):
        pyexcel.save_as(array=dataset, dest_file_name=str(resultset))

    elif str(resultset).endswith('.xlsx'):
        pyexcel.save_as(array=dataset, dest_file_name=str(resultset))

    elif str(resultset).endswith('.ods'):
        print(f'{BCOLORS["WARNING"]}NOTE: The writing of the ODS file takes some time. Your terminal did not crash. Please wait ...{BCOLORS["ENDC"]}')
        pyexcel.save_as(array=dataset, dest_file_name=str(resultset))

    print('Wrote ' + str(counter) + ' lines')


whatsapp-converter.py
'''whatsapp-converter

Use this application to convert your exported WhatsApp chat to a CSV or ODS file.

1. The conversion is done locally. No data is shared with the internet!
2. The resultset is a CSV file which you can import into your favorite calculation application such as Excel or LibreOffice. You can also directly export to a LibreOffice spreadsheet file.

'''
import io
import re
import sys
import argparse
import datetime
from datetime import date
import xlwt
from tqdm import tqdm

Constant = {
    "VERSION": 'v0.3.7'
}

# TODO Add colors to output
# bcolors = {
#     "HEADER": "\033[95m",
#     "OKBLUE": "\033[94m",
#     "OKGREEN": "\033[92m",
#     "WARNING": "\033[93m",
#     "FAIL": "\033[91m",
#     "ENDC": "\033[0m",
#     "BOLD": "\033[1m",
#     "UNDERLINE": "\033[4m"
# }

lastentry = {
    "lastdate": date.today(),
    "lasttime": date.today(),
    "lastname": ""
}

def parse(line, local_args):
    '''Parse each line

    line = Single line to parse
    args = Arguments from the command line. In this case only verbose and debug switches.
    '''
    patterndate = "^((\d{1,2})([\/|\.])(\d{1,2})[\/|\.](\d{1,2}))\,\ (\d{1,2}:\d{1,2})(?::\d{1,2})?\ ?(AM|PM|am|pm)?([\:\ |\ \-\ ][^:])(.+?)\:"
    pattern = "^((\d{1,2})([\/|\.])(\d{1,2})[\/|\.](\d{1,2}))\,\ (\d{1,2}:\d{1,2})(?::\d{1,2})?\ ?(AM|PM|am|pm)?([\:\ |\ \-\ ][^:])(.+?)\:\ (.*)"
    found = False
    dataset = ['empty', '', '', '', '', '']
    # if verbose: print(prefix + line)
    # if verbose: print(BColors.FAIL + prefix + line)
    # Identify the date format in the chat line

    if re.match(re.compile(patterndate, re.VERBOSE), line):
        found = True

    elif re.match(re.compile(r"^[\t ]*\n", re.VERBOSE), line):
        # Empty line
        if local_args.debug:
            print("Empty line removed")

    else:
        if local_args.debug:
            print("Appending line found")

        newline = (re.match(re.compile(r"^(.*)", re.VERBOSE), line))

        # Create the dataset if commandline argument was to create a new line
        # TODO if (local_args.newline):
        # if 1:
        #     dataset = ['new', str(lastentry["lastdate"]), str(lastentry["lasttime"]), lastentry["lastname"], newline.group(0)]
        dataset = ['new', str(lastentry["lastdate"]), str(lastentry["lasttime"]), lastentry["lastname"], newline.group(0)]
        #     # if local_args.verbose | local_args.debug:
        #           print(dataset)
        #
        # else:
        #     # Otherwise make sure it is appended to the existing line
        #     dataset = ['append', '', '', '', newline.group(0)]
        #     if local_args.verbose | local_args.debug:
        #          print(dataset)

    if found:
        # Make the match, assign to the groups
        match = re.match(re.compile(pattern, re.VERBOSE), line)

        if match and match.group(9) != 'M':
            # 21/12/19 Date Format
            if match.group(3) == '/' and match.group(8) == ': ':
                date = datetime.datetime.strptime(match.group(1), "%d/%m/%y").date()
            # 12/21/19 Date Format
            elif match.group(3) == '/' and (match.group(8) == ' -' or match.group(8) == '- '):
                date = datetime.datetime.strptime(match.group(1), "%m/%d/%y").date()
            # 21.12.19 Date Format
            else:
                date = datetime.datetime.strptime(match.group(1), "%d.%m.%y").date()

            if match.group(7):
                time = datetime.datetime.strptime(match.group(6) + " " + match.group(7), "%I:%M %p").time()
            else:
                time = datetime.datetime.strptime(match.group(6), "%H:%M").time()

            # Buffer date, time, name for next line messages
            lastentry["lastdate"] = date.strftime("%Y-%m-%d")
            lastentry["lasttime"] = time.strftime("%H:%M")
            lastentry["lastname"] = match.group(9)

            # Create the dataset for the new message
            dataset = ['new', date.strftime("%Y-%m-%d"), time.strftime("%H:%M"), str(match.group(9)), match.group(10)]
            if local_args.verbose | local_args.debug:
                print(dataset)

    return dataset

def convert(local_args):
    '''Convert the input file

    1. Open the file
    2. Loop to input line by line and send it to parser
    3. Close the file

    '''
    # Store the number of lines of the input file
    line_count = 0

    if local_args.verbose:
        print("Verbosity turned on")

    if local_args.debug:
        print("Debug turned on")

    try:
        # Count the number of total lines
        with io.open(local_args.filename, "r", encoding="utf-8") as file:
            for line in file:
                # if line.strip():
                line_count += 1

        # Convert lines to csv
        with io.open(local_args.filename, "r", encoding="utf-8") as file:
            content = file.readlines()

    except IOError as e:
        print("File \"" + local_args.filename + "\" cannot be found.")
        sys.exit()

    print("Converting data now")

    # Count number of chatlines without empty lines
    counter = 0

    if local_args.debug:
        print('Open export file ' + local_args.resultset)

    # Define export file header
    header = ['Date and Time', 'Date', 'Time', 'Name', 'Message']

    # Select export formats
    if str(local_args.resultset).endswith('.csv'):

        # Open result filename
        csv = io.open(local_args.resultset, "w", encoding="utf-8")

        # Write headers
        csv.write(header[0] + '|' + header[1] + '|' + header[2] + '|' + header[3] + '|' + header[4] + '|' + '\n')

    elif str(local_args.resultset).endswith('.ods'):
        workbook = xlwt.Workbook()
        worksheet = workbook.add_sheet(local_args.filename)

        worksheet.write(0, 0, header[0])
        worksheet.write(0, 1, header[1])
        worksheet.write(0, 2, header[2])
        worksheet.write(0, 3, header[3])
        worksheet.write(0, 4, header[4])

        ws_counter = 1

    # TODO Append line with buffer before writing
    # Show progress via tqdm
    for line in tqdm(content, total=line_count, ncols=120):
        if local_args.debug and line == '':
            print(line)

        dataset = parse(line, local_args)

        if dataset[0] != 'empty':

            # Write to .csv file
            if str(local_args.resultset).endswith('.csv'):
                csv.write(dataset[1] + ' ' + dataset[2] + '|' + dataset[1] + '|' + dataset[2] + '|' + dataset[3] + '|' + dataset[4] + '\n')

            # Write to .ods file
            elif str(local_args.resultset).endswith('.ods'):
                worksheet.write(ws_counter, 0, counter)
                worksheet.write(ws_counter, 1, dataset[1] + ' ' + dataset[2])
                worksheet.write(ws_counter, 2, dataset[2])
                worksheet.write(ws_counter, 3, dataset[3])
                worksheet.write(ws_counter, 4, dataset[4])
                ws_counter += 1

        # Print progress
        if line.strip():
            counter += 1
            # print('Wrote ' + str(counter) + ' lines of ' + str(line_count) + ' lines', end='\r')

    print('Wrote ' + str(counter) + ' lines')

    # Close the resultfiles
    if str(local_args.resultset).endswith('.csv'):
        csv.close()

    elif str(local_args.resultset).endswith('.ods'):
        workbook.save('resultset.ods')

if __name__ == "__main__":
    parser = argparse.ArgumentParser(prog='whatsapp-converter', epilog='For reporting bugs or requesting features, please visit https://github.com/sandsturm/whatsapp-converter/ and create an issue')
    parser.add_argument('filename', metavar='filename', type=str, help='the WhatsApp file containing the exported chat')
    parser.add_argument('resultset', default='resultset.csv', nargs='?', help='filename of the resultset, default resultset.csv. Use .csv to write a comma separated file. Use .ods to write to a LibreOffice spreadsheet file')
    parser.add_argument('-v', '--verbose', help='increase output verbosity', action='store_true')
    parser.add_argument('-d', '--debug', help='increase output verbosity to debug', action='store_true')

    # parser.add_argument("-nl", "--newline", help="message across various lines is counted as a new message", action="store_true")
    args = parser.parse_args()

    convert(args)
